src/libraries/System.Net.Security/src/System/Net/Security/SslSessionsCache.cs

```diff
@@ -191,20 +198,13 @@ internal static void CacheCredential(SafeFreeCredentials creds, byte[]? thumbPri
 
                             for (int i = 0; i < toRemoveAttempt.Length; ++i)
                             {
-                                cached = toRemoveAttempt[i].Value;
+                                var freeCreds = toRemoveAttempt[i].Value.Target;
 
-                                if (cached != null)
+                                if (freeCreds == null || freeCreds.IsClosed || freeCreds.IsInvalid)
                                 {
-                                    creds = cached.Target;
-                                    cached.Dispose();
-
-                                    if (!creds.IsClosed && !creds.IsInvalid && (cached = SafeCredentialReference.CreateReference(creds)) != null)
```

Comment:
Do you know what the purpose of this `cached = SafeCredentialReference.CreateReference(creds)` was?  This is removing it and the adding of it back into the table.

Reply:
The logic was:
1) dispose cahced - cached.Target becomes null, the SafeCredentialReference ref count decrease
2) if creds are still valid, create SafeCredentialReference from it (cached should not be null as CreateReference returns null when creds.IsInvalid )
3) if the SafeCredentialReference is successfully created (the ref count increase), it replace the disposed one in the cache.
- if the creds is valid, the step 1-3 is noop
- if creds is invalid, it is removed from the s_cachedCreds

changed to:
1) if creds are invalid, and are removed from cache, then dispose SafeCredentialReference wrapper which decrease ref count

Reply:
There is a subtle difference between the old logic and the new logic, and I expect it's why it was written the way it was written.

When you Dispose of a SafeHandle and there's an outstanding ref count on it, the disposal is delayed until that outstanding ref count is released.  So imagine there was a handle in the collection that had Dispose called on it.  That Dispose won't have taken effect yet, because there's an outstanding ref count from this SafeCredentialReference.  The previous logic was disposing of the SafeCredentialReference, which would have allowed an outstanding Dispose to take effect, after which IsClosed might be true.  But in the new logic, the SafeCredentialReference isn't being disposed unless the handle is already disposed, and as such, it's going to miss removing handles that were disposed, because the disposal couldn't have taken effect yet.

This means that the new code is no longer removing disposed handles, because in the common case disposal won't have actually taken effect because of the SafeCredentialReference.  Given that, I expect the new logic is broken.

Reply:
I reverted the old cleanup logic. Thanks for catching this bug